File: Examples/CSharp/x-IMU3-Examples/Examples/SerialDiscovery.cs
using System;

namespace Ximu3Examples
{
    class SerialDiscovery
    {
        public SerialDiscovery()
        {
            if (Helpers.YesOrNo("Use async implementation?") == true)
            {
                Ximu3.SerialDiscovery discovery = new Ximu3.SerialDiscovery(SerialDiscoveryEvent);
                Helpers.Wait(-1);
                discovery.Dispose();
            }
            else
            {
                Ximu3.DiscoveredSerialDevice[] devices = Ximu3.SerialDiscovery.Scan(2000);
                Console.WriteLine("Devices updated (" + devices.Length + " devices available)");
                PrintDevices(devices);
            }
        }

        private void SerialDiscoveryEvent(Object sender, Ximu3.SerialDiscoveryEventArgs args)
        {
            Console.WriteLine("Discovered " + args.devices.Length + " devices");
            PrintDevices(args.devices);
        }

        private void PrintDevices(Ximu3.DiscoveredSerialDevice[] devices)
        {
            foreach (Ximu3.DiscoveredSerialDevice device in devices)
            {
                Console.WriteLine(device.DeviceName + " - " + device.SerialNumber + ", " + device.ConnectionInfo);
                // Console.WriteLine(device); // alternative to above
            }
        }
    }
}

File: Examples/Python/serial_discovery.py
import helpers
import ximu3


def print_devices(devices):
    for device in devices:
        print(device.device_name + " - " +
              device.serial_number + ", " +
              device.connection_info.to_string())
        # print(device.to_string())  # alternative to above


def callback(devices):
    print("Devices updated (" + str(len(devices)) + " devices available)")
    print_devices(devices)


if helpers.yes_or_no("Use async implementation?"):
    _ = ximu3.SerialDiscovery(callback)
    helpers.wait(-1)
else:
    devices = ximu3.SerialDiscovery.scan(2000)
    print("Discovered " + str(len(devices)) + " devices")
    print_devices(devices)
